### ISSUE TITLE
Specific Code Line Highlighting

### ISSUE THREAD

**Issue by aalvrz** [rouge-ruby/rouge#617]
I wanted to know if it's possible to highlight specific lines of code? For example:

```ruby
def add(a,b)
  c = a + b
  puts c # I want to highlight this line
end
```

**NOTE:** By *highlight* I do not mean syntax highlighting, but rather line highlighting (some applying of background color to the specified line in the code block).

**Comment by revathskumar** [comment]
In this case I tried to use `Diff` highlighting, but AFAIK by default rouge doesn't support missing of syntax highlighting of diff and other language. I raised a issue #642 related to this. 

Later as my friend suggested, wrote a custom lexer by extending the `JSON` lexer and prepended the `Diff` lexer rules, I was able to make it work. See the issue #642 for code example.

Also here is my [example workaround](https://github.com/revathskumar/rouge-json-diff-lexer-example)

**Comment by BigChief45** [comment]
@revathskumar That issue seems to be similar to what I am trying to achieve, except I am not concerned with diffs. I just want Rouge to somehow be able to identify some characters that indicate the highlighting of a specific line in a code block.

I am guessing this must be added to the main lexer (or something like that). I could try working on it for a PR but some help would be appreciated.

**Comment by jneen** [comment]
I have a spec I'm working on for some kind of "HLML" markup to allow arbitrary combinations of highlighters and custom tokens, which I think is what would be needed here. However, I don't have the bandwidth to work on it any time soon - but I would be happy to pass of a sketch of it if one of yall wanted to take it on.

**Comment by BigChief45** [comment]
@jneen That would be great. I would like giving it a try.

**Comment by BigChief45** [comment]
@jneen Any updates on this?

**Comment by BigChief45** [comment]
@gfx @dblessing

Any updates on this?

**Comment by MischaDeNola** [comment]
@jneen Hi, I'm a writer, not a developer, so my apologies for sounding like an idiot:

A couple years ago I wrote a custom lexer for rouge for my company's own proprietary language. I recently added a diff rule (for strong) per @revathskumar's example so that we could draw attention to particular lines in code blocks when writing instructional docs. 

To keep customers from copying the `!` from our code blocks,  I split the labeling with the **groups** thing:

```ruby
        rule /(^!)(.*$\n?)/ do
          groups Generic::Subheading, Generic::Strong
        end
```
I then just set the style for **Subheading** to `display: none;` and it worked like a charm. We highlight the line while excluding the bang. Yay!

However, I recently discovered a bug with this in a couple of rare situations where the rule would trigger inappropriately--not respecting the start-of-line assertion.

```
function(
domain!whatever(
string: "hello"
)
)
```

Would render that as:

> function(
> domain**whatever(**
> string: "hello"
> )
> )

When I removed the groups, it did NOT trigger on `domain!whatever`:

```ruby
rule /^!.*$\n?/, Generic::Strong
```

I don't know why using **groups** makes it act like this. I'm working around the problem by just adding a rule to catch `domain!whatever`.

```diff
        rule /(^!)(.*$\n?)/ do
          groups Generic::Subheading, Generic::Strong
        end
+        rule /(domain!\[a-zA-Z_][a-zA-Z0-9_]*)/, Str # this stops weird bug
```

Just in case this is an actual bug, I'm letting you know.

Thank you for your time (and for rouge!)

**Comment by jneen** [comment]
@MischaDeNola happy to help, but could you move this to a different issue?

**Comment by MischaDeNola** [comment]
Done.

**Comment by pyrmont** [comment]
Hi everyone :) @jneen added a special [escape lexer](https://github.com/rouge-ruby/rouge/pull/1152) recently that I _think_ might allow this using HTML. I'd suggest checking that out :)

**Comment by pyrmont** [comment]
After thinking about this again, I don't think the escape lexer does solve the issue. I'll add it back to the pile. Apologies for the delay :(

**Comment by ernstki** [comment]
I hope stalebot doesn't re-open this on my account (I dunno how that works), but it looks like this was implemented in #264, right?

**Comment by woodjosj** [comment]
Hey
Sent from Yahoo Mail on Android 
 
  On Fri, Aug 13, 2021 at 11:09 PM, Kevin ***@***.***> wrote:   


I hope stalebot doesn't re-open this on my account (I dunno how that works), but it looks like this was implemented in #264, right?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
